src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,138 @@ public void GenerateHeader()
             PopBlock();
         }
 
+        private class DelegateSignature
+        {
+            public string retType, signature;
+            public bool suppressWarning;
+            public System.Runtime.InteropServices.CallingConvention callingConvention;
+
+            public override int GetHashCode()
+            {
+                return retType.GetHashCode() ^ signature.GetHashCode() ^ callingConvention.GetHashCode();
+            }
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as DelegateSignature);
+            }
+            public bool Equals(DelegateSignature obj)
+            {
+                return obj != null && CompareWith(obj);
+            }
+            private bool CompareWith(DelegateSignature ds)
+            {
+                if (!retType.Equals(ds.retType) || !signature.Equals(ds.signature) || suppressWarning != ds.suppressWarning ||
+                    callingConvention != ds.callingConvention)
+                {
+                    return false;
+                }
+                return true;
+            }
+        }
+
+        private string GenerateDelegateSignature(Function function, out CSharpTypePrinterResult retType)
+        {
+            var @params = new List<string>();
```

Comment:
All of this code seems an identical copy of GatherInternalParams (https://github.com/mono/CppSharp/blob/master/src/Generator/Generators/CSharp/CSharpTextTemplate.cs#L600). Please reuse that method instead of repeating the code. In case there are some minor differences, introduce a parameter.